File: AC/library/ilib/machine-audio/machine-audio.py
# AC ilib: machine-audio — Speech-to-Text and Text-to-Speech (Python backend)
import subprocess, shutil, tempfile, os

def maudio_speak(text, rate=175):
    """Speak text aloud using eSpeak-NG or system TTS."""
    if shutil.which("espeak-ng"):
        subprocess.run(["espeak-ng", "-s", str(rate), text], check=False)
        return True
    elif shutil.which("say"):  # macOS
        subprocess.run(["say", text], check=False)
        return True
    elif shutil.which("espeak"):
        subprocess.run(["espeak", "-s", str(rate), text], check=False)
        return True
    elif shutil.which("spd-say"):
        subprocess.run(["spd-say", "-r", str(rate - 175), text], check=False)
        return True
    else:
        print(f"[machine-audio::speak] {text}")
        return False

def maudio_speak_async(text, rate=175):
    """Speak text without blocking."""
    if shutil.which("espeak-ng"):
        subprocess.Popen(["espeak-ng", "-s", str(rate), text])
    elif shutil.which("say"):
        subprocess.Popen(["say", text])
    elif shutil.which("spd-say"):
        subprocess.Popen(["spd-say", "-r", str(rate - 175), text])
    else:
        print(f"[machine-audio::speak_async] {text}")

def maudio_listen(timeout_ms=5000):
    """Record from microphone and return transcribed text.
    Requires: arecord (Linux) + whisper CLI, or SpeechRecognition package."""
    secs = max(1, timeout_ms // 1000)
    # Try SpeechRecognition (pip install SpeechRecognition pyaudio)
    try:
        import speech_recognition as sr
        r = sr.Recognizer()
        with sr.Microphone() as src:
            r.adjust_for_ambient_noise(src, duration=0.3)
            audio = r.listen(src, timeout=secs)
        return r.recognize_google(audio)
    except Exception:
        pass
    # Fallback: arecord + whisper CLI
    tmp = tempfile.mktemp(suffix=".wav")
    try:
        subprocess.run(["arecord", "-d", str(secs), "-f", "cd", "-t", "wav", tmp],
                       check=False, capture_output=True)
        out = subprocess.run(["whisper", "--model", "tiny", tmp,
                              "--output-txt", "--output-file", tmp + "_out"],
                             capture_output=True, check=False)
        txt_path = tmp + "_out.txt"
        if os.path.exists(txt_path):
            with open(txt_path) as f: return f.read().strip()
    finally:
        for p in [tmp, tmp + "_out.txt"]:
            try: os.remove(p)
            except: pass
    return ""

def maudio_tts_ok():
    return bool(shutil.which("espeak-ng") or shutil.which("say") or shutil.which("espeak") or shutil.which("spd-say"))

def maudio_stt_ok():
    try:
        import speech_recognition; return True
    except ImportError: pass
    return bool(shutil.which("whisper"))

File: library/ilib/machine-audio/ffi/MachineAudio.java
// AC ilib: machine-audio — Speech-to-Text and Text-to-Speech (Java)
import java.io.*;
import java.nio.file.*;

public class MachineAudio {

    /**
     * maudio_speak - Speak text aloud
     * @param text Text to speak
     * @param rate Speech rate (50-300, default 175)
     * @return true if successful
     */
    public static boolean maudioSpeak(String text, int rate) {
        if (text == null || text.isEmpty()) return false;
        rate = Math.max(50, Math.min(300, rate > 0 ? rate : 175));

        try {
            // Try espeak-ng
            if (commandExists("espeak-ng")) {
                ProcessBuilder pb = new ProcessBuilder("espeak-ng", "-s", String.valueOf(rate), text);
                Process p = pb.start();
                p.waitFor();
                return true;
            }
            // Try say (macOS)
            if (commandExists("say")) {
                ProcessBuilder pb = new ProcessBuilder("say", text);
                Process p = pb.start();
                p.waitFor();
                return true;
            }
            // Try espeak
            if (commandExists("espeak")) {
                ProcessBuilder pb = new ProcessBuilder("espeak", "-s", String.valueOf(rate), text);
                Process p = pb.start();
                p.waitFor();
                return true;
            }
            // Try spd-say
            if (commandExists("spd-say")) {
                ProcessBuilder pb = new ProcessBuilder("spd-say", "-r", String.valueOf(rate - 175), text);
                Process p = pb.start();
                p.waitFor();
                return true;
            }

            System.out.println("[machine-audio::speak] " + text);
            return false;
        } catch (Exception e) {
            System.out.println("[machine-audio::speak] " + text);
            return false;
        }
    }

    /**
     * maudio_speak - Speak text aloud (default rate)
     */
    public static boolean maudioSpeak(String text) {
        return maudioSpeak(text, 175);
    }

    /**
     * maudio_speak_async - Speak text without blocking
     */
    public static void maudioSpeakAsync(String text, int rate) {
        if (text == null || text.isEmpty()) return;
        rate = Math.max(50, Math.min(300, rate > 0 ? rate : 175));

        final int finalRate = rate;
        final String finalText = text;

        new Thread(() -> {
            try {
                if (commandExists("espeak-ng")) {
                    new ProcessBuilder("espeak-ng", "-s", String.valueOf(finalRate), finalText).start();
                } else if (commandExists("say")) {
                    new ProcessBuilder("say", finalText).start();
                } else if (commandExists("espeak")) {
                    new ProcessBuilder("espeak", "-s", String.valueOf(finalRate), finalText).start();
                } else if (commandExists("spd-say")) {
                    new ProcessBuilder("spd-say", "-r", String.valueOf(finalRate - 175), finalText).start();
                } else {
                    System.out.println("[machine-audio::speak_async] " + finalText);
                }
            } catch (IOException e) {
                System.out.println("[machine-audio::speak_async] " + finalText);
            }
        }).start();
    }

    /**
     * maudio_speak_async - Speak text without blocking (default rate)
     */
    public static void maudioSpeakAsync(String text) {
        maudioSpeakAsync(text, 175);
    }

    /**
     * maudio_listen - Record from microphone and transcribe
     * @param timeoutMs Timeout in milliseconds
     * @return Transcribed text
     */
    public static String maudioListen(int timeoutMs) {
        timeoutMs = Math.max(1000, timeoutMs > 0 ? timeoutMs : 5000);
        int secs = (timeoutMs + 999) / 1000;

        try {
            String tmpFile = Files.createTempFile("maudio_", ".wav").toString();
            String outFile = tmpFile + "_out";

            // Record audio
            boolean recorded = false;
            if (commandExists("arecord")) {
                ProcessBuilder pb = new ProcessBuilder(
                    "arecord", "-d", String.valueOf(secs), "-f", "cd", "-t", "wav", tmpFile
                );
                Process p = pb.start();
                p.waitFor();
                recorded = Files.exists(Paths.get(tmpFile));
            } else if (commandExists("ffmpeg")) {
                ProcessBuilder pb = new ProcessBuilder(
                    "ffmpeg", "-f", "avfoundation", "-i", ":0", "-t", String.valueOf(secs),
                    tmpFile, "-y"
                );
                Process p = pb.start();
                p.waitFor();
                recorded = Files.exists(Paths.get(tmpFile));
            }

            if (!recorded) return "";

            // Transcribe with whisper
            if (commandExists("whisper")) {
                ProcessBuilder pb = new ProcessBuilder(
                    "whisper", "--model", "tiny", tmpFile,
                    "--output-txt", "--output-file", outFile
                );
                Process p = pb.start();
                p.waitFor();

                String txtPath = outFile + ".txt";
                if (Files.exists(Paths.get(txtPath))) {
                    String text = new String(Files.readAllBytes(Paths.get(txtPath))).trim();
                    // Cleanup
                    try {
                        Files.deleteIfExists(Paths.get(tmpFile));
                        Files.deleteIfExists(Paths.get(txtPath));
                    } catch (Exception e) {}
                    return text;
                }
            }

            // Cleanup on failure
            try { Files.deleteIfExists(Paths.get(tmpFile)); } catch (Exception e) {}
            return "";
        } catch (Exception e) {
            return "";
        }
    }

    /**
     * maudio_tts_ok - Check if Text-to-Speech is available
     * @return 1 if available, 0 otherwise
     */
    public static int maudioTtsOk() {
        return (commandExists("espeak-ng") || commandExists("say") ||
                commandExists("espeak") || commandExists("spd-say")) ? 1 : 0;
    }

    /**
     * maudio_stt_ok - Check if Speech-to-Text is available
     * @return 1 if available, 0 otherwise
     */
    public static int maudioSttOk() {
        return commandExists("whisper") ? 1 : 0;
    }

    /**
     * Helper: Check if command exists in PATH
     */
    private static boolean commandExists(String cmd) {
        String os = System.getProperty("os.name").toLowerCase();
        String[] cmd_array;

        try {
            if (os.contains("win")) {
                cmd_array = new String[]{"cmd.exe", "/c", "where " + cmd};
            } else {
                cmd_array = new String[]{"/bin/sh", "-c", "which " + cmd};
            }

            Process p = Runtime.getRuntime().exec(cmd_array);
            return p.waitFor() == 0;
        } catch (Exception e) {
            return false;
        }
    }
}
